homeassistant/components/mqtt/text.py

```diff
@@ -0,0 +1,240 @@
+        )
+
+        self._command_template = MqttCommandTemplate(
+            config.get(CONF_COMMAND_TEMPLATE),
+            entity=self,
+        ).async_render
+        self._value_template = MqttValueTemplate(
+            config.get(CONF_VALUE_TEMPLATE),
+            entity=self,
+        ).async_render_with_possible_json_value
+        optimistic: bool = config[CONF_OPTIMISTIC]
+        self._optimistic = optimistic or config.get(CONF_STATE_TOPIC) is None
+
+    def _validate_text_value(self, payload: str, topic: str) -> str | None:
+        """Return an error string if a text value is invalid."""
+        if len(payload) > self._attr_native_max:
+            return f"'{topic}', size > max size ({self._attr_native_max})"
+
+        if len(payload) < self._attr_native_min:
+            return f"'{topic}', size < min size ({self._attr_native_min})"
+
+        if self._compiled_pattern and not self._compiled_pattern.match(payload):
+            return f"'{topic}', text does not match pattern '{self._attr_pattern}'"
+
+        return None
+
+    def _prepare_subscribe_topics(self) -> None:
+        """(Re)Subscribe to topics."""
+        topics: dict[str, Any] = {}
+
+        def add_subscription(
+            topics: dict[str, Any], topic: str, msg_callback: Callable
+        ) -> None:
+            if self._config.get(topic) is not None:
+                topics[topic] = {
+                    "topic": self._config[topic],
+                    "msg_callback": msg_callback,
+                    "qos": self._config[CONF_QOS],
+                    "encoding": self._config[CONF_ENCODING] or None,
+                }
+
+        @callback
+        @log_messages(self.hass, self.entity_id)
+        def handle_state_message_received(msg: ReceiveMessage) -> None:
+            """Handle receiving state message via MQTT."""
+            payload = str(self._value_template(msg.payload))
+            if error := self._validate_text_value(payload, msg.topic):
+                _LOGGER.warning("Ignoring text update from %s", error)
+
+            self._attr_native_value = payload
+            get_mqtt_data(self.hass).state_write_requests.write_state_request(self)
+
+        add_subscription(topics, CONF_STATE_TOPIC, handle_state_message_received)
+
+        self._sub_state = subscription.async_prepare_subscribe_topics(
+            self.hass, self._sub_state, topics
+        )
+
+    async def _subscribe_topics(self) -> None:
+        """(Re)Subscribe to topics."""
+        await subscription.async_subscribe_topics(self.hass, self._sub_state)
+
+    @property
+    def assumed_state(self) -> bool:
+        """Return true if we do optimistic updates."""
+        return self._optimistic
+
+    async def async_set_value(self, value: str) -> None:
+        """Change the text."""
+        if CONF_COMMAND_TOPIC not in self._config:
+            raise ValueError("Command topic not set, cannot set the text value")
```

Comment:
Maybe a command topic should be required if `TextEntity` doesn't have a read only flag? A read only `TextEntity` would be better represented as a sensor?